### ISSUE TITLE
-reindex fails if blk0*.dat files read-only

### ISSUE THREAD

**Issue by luke-jr** [bitcoin/bitcoin#2039]
To reproduce: chmod -r datadir/blk0_.dat
Expected behaviour: Reindex read-only block file, and create a new blk0_.dat file when there are new blocks to add.

**Comment by laanwj** [comment]
bitcoind makes the (generally valid) assumption a that it can write anything in its data directory. Reindexing from read-only block files may be a useful feature in some cases, but I'm removing the bug label.

**Comment by mruddy** [comment]
I was just looking at this. The fix introduced by https://github.com/bitcoin/bitcoin/pull/24858 gets us very close to being able to allow read only blk files that are already finalized. This only makes sense with **finalized** block files as blk files that have logical space under the 128 MiB limit need to be writable for appending additional blocks. Even with the #24858 fix, `BlockManager::FlushBlockFile` still tries to open the finalized block file for writing. But, I don't think it would be very difficult to keep track in `BlockManager::FindBlockPos` via the `m_blockfile_info[nFile]` to know if the current blk file has been appended. We'd have to add another data member to track if any new blocks have been appended to the blockfile. If it hasn't been appended, then we could accept the blk flush failing as it was not changed, and the failure was caused by not being able to open it for writing.

For reference, I tested current master branch functionality and got the following while using only a single finalized blk00000.dat (fe4b764e0ce523cfce6d9e4f326e0d0977ee317a51b4a129cda707cfa4699dd7  blocks/blk00000.dat):
```
Unable to open file /tmp/btc/blocks/blk00000.dat
ERROR: Flush: failed to open file 0
*** Flushing block file to disk failed. This is likely the result of an I/O error.
Error: A fatal internal error occurred, see debug.log for details
```
